#load libraries needed
library(phytools)

#create this file only the first time then we append to it
pic.output = data.frame("Immune_Trait", "Diversity_Trait", "Marker_Gene_Name", "N_Microbiomes", "N_Genomes", "Equation_Zero_Inc", "pVal_Zero_Inc", "rSquared_Zero_Inc", "Equation_Nonzero", "pVal_Nonzero", "rSquared_Nonzero")
pic.output
write.table(pic.output,file="symbiodinium_pics/PIC_Sym_output_its2&clade.csv",append=TRUE, sep=",", row.names=FALSE, col.names=FALSE)

#Load the trait table and tree for the analysis
#There are 2 trait tables: one for its2 input/symbiodinium_trait_table_its2.tsv
#one for clade: "input/symbiodinium_trait_table_clade.tsv"
trait_table <-"input/symbiodinium_trait_table_clade.tsv"
tree <- "input/symbiodinium_phylogeny_muscle_gblocks_phyml_via_phylogenyfr_09172021.newick"
#Pick the columns to analyze
#X traits can be: TIR_total, TIR_total_unique, LRR_total, LRR_total_unique
x_trait <-"Lectin_total"
#Y traits can be: obs_asvs, ln_asvs, dominance, gini_index, simpson_e, faith_pd
y_trait <-"faith_pd"

print(paste("Analyzing",x_trait,"vs.",y_trait))

trait_table <- read.table(trait_table,comment.char="",header=T,row.names=1,as.is=T)
tree <- read.tree(tree)
#print(trait_table)

#Filter table to tree tips
trait_table <- trait_table[rownames(trait_table) %in% tree$tip.label,]

#Coerce trait columns to numeric, possibly inducing NAs
trait_table[,x_trait]<-as.numeric(trait_table[,x_trait])
trait_table[,y_trait]<-as.numeric(trait_table[,y_trait])

#Drop NA rows for our x and y traits
trait_table<- trait_table[!is.na(trait_table[,x_trait]),]
trait_table <- trait_table[!is.na(trait_table[,y_trait]),]

#Restrict to just our x column of interest
X <- trait_table[,x_trait]
names(X) <- rownames(trait_table)

#Get our y-value column of interest
Y <- trait_table[,y_trait]
names(Y) <- rownames(trait_table)



#Ensure tree has been filtered so all tips are names in X and Y
filtered_tree <- drop.tip(tree,tree$tip.label[!tree$tip.label %in% names(Y)])
tree <- drop.tip(filtered_tree,tree$tip.label[!tree$tip.label %in% names(X)])

#Dichotomize tree
tree <- multi2di(tree)


#Check tree structure
print(paste("Binary tree:", is.binary(tree)))
min_branch_length = 0.0024
#min_branch_length = 0.000001
tree$edge.length[tree$edge.length==0] <- min_branch_length
na_edges <- tree$edge[is.na(tree$edge.length[tree$edge[,2]])]
print(na_edges)
tree$edge.length[na_edges] <- min_branch_length
print(tree$edge.length[na_edges])
tree$edge.length[is.na(tree$edge.length[tree$edge[,2]])] <- min_branch_length
#print the edge lengths for the first runs
#print(paste("Edge length list:",tree$edge.length[tree$edge[,2]]))
#plot(tree)

print(paste("Fitting raw linear regression before PIC for",x_trait,"(x) vs. ",y_trait,"(y)"))
#Note for raw regression we don't force the regression through the origin
raw_fitYX <-lm(Y ~ X)

print(summary(raw_fitYX))

#print("Calculating PIC")
#****POSITIVIZE PIC VALUES (negative signs are arbitrary)
#NOTE: this is not just the absolute value since the difference x=-1 y=2
#has a different interpretation than x=1 y=2

#Code snippet from:https://github.com/bomeara/ComparativeMethodsInR/blob/master/ContinuousTrait_Answers.R
#for contrasts, you should positivize them, since the order doesn't matter. This is NOT taking absolute value.
PositivizeContrasts <- function(x, y) {
  #Cache the sign of x so it doesn't
  #change as we reflect points about x-axis!
  sign_of_x <- sign(x)
  x.positivized <- x * sign_of_x
  y.positivized <- y * sign_of_x
  return(cbind(x.positivized, y.positivized))
}

raw.pic.X <- pic(X,tree)
raw.pic.Y <- pic(Y,tree)
#print(raw.pic.X)
#print(raw.pic.Y)

positivized.results <- PositivizeContrasts(raw.pic.X, raw.pic.Y)
pic.X <-positivized.results[,1]
pic.Y <-positivized.results[,2]



for (method in c('spearman','pearson')){
  corYX <- cor.test(pic.X, pic.Y, method=method, exact=FALSE)
  print(paste("Correlation test for PICs of",y_trait,"(y) vs. ",x_trait,"(x)"))
  
  print(paste("Summary cor.test for PICs ",method))
  print(corYX)
}

#Try to plot PIC omitting zero y-values (generated by species --> sample mappings)
epsilonX = 0.05 * min(X) + 0.00000001
epsilonY = 0.05 * min(Y) + 0.00000001
Y_nonzero_indices = which(abs(pic.Y) > epsilonY)
pic.X.nonzero = pic.X[Y_nonzero_indices]
pic.Y.nonzero = pic.Y[Y_nonzero_indices]

X_nonzero_indices = which(abs(pic.X) > epsilonX)
pic.X.nonzero = pic.X.nonzero[X_nonzero_indices]
pic.Y.nonzero = pic.Y.nonzero[X_nonzero_indices]





# regress through origin. Somewhat debatable if  we want proportion of variance around 0
#or the mean contrast (because expected contrast is 0..but we are removing zeros).
#I think calculating the proportion of variance around the mean (-1) is safer.

fitYX <- lm(pic.Y ~ pic.X -1 )
print(paste("Summary lm pic.Y ~ pic.X +0 INCLUDING zero change PICs for" ,x_trait,"(x) vs. ",y_trait,"(y)"))
print(summary(fitYX))
#extract the r2,p value, and model equation
rSquared_zero_inc <- summary(fitYX)$r.squared
pVal_zero_inc <- anova(fitYX)$'Pr(>F)'[1]
model_zero_inc <- "lm(pic.Y ~ pic.X -1 )"

#print out values that will go in the final table
print(model_zero_inc)
print(pVal_zero_inc)
print(rSquared_zero_inc)


fitYX.nonzero <- lm(pic.Y.nonzero ~  pic.X.nonzero -1 ) # both regressions
print(paste("Summary lm pic.Y ~ pic.X +0 non-zero PICs only for" ,x_trait,"(x) vs. ",y_trait,"(y)"))
print(summary(fitYX.nonzero))
#extract the r2, p value, and model equation
rSquared_nonzero <-summary(fitYX.nonzero)$r.squared
pval_nonzero <- anova(fitYX.nonzero)$'Pr(>F)'[1]
model_nonzero <- "lm(pic.Y.nonzero ~ pic.X.nonzero -1)"

#print out values that will go in the final table
print(model_nonzero)
print(pval_nonzero)
print(rSquared_nonzero)

#write results from the PICS (zeros and nonzeros) to the created csv file

trait_table.df <- data.frame(trait_table)
#this extracts the marker gene name (ITS2 or clade)
marker_gene <- trait_table.df$Phylo_Classification[1]
print(marker_gene)
n_microbe <- sum(trait_table.df$Num_samples_microbe)
n_genome <- sum(trait_table.df$Num_samples_genome)
output_row <- data.frame(x_trait, y_trait, marker_gene, n_microbe, n_genome, model_zero_inc, pVal_zero_inc, rSquared_zero_inc, model_nonzero, pval_nonzero, rSquared_nonzero )
print(output_row)
write.table(output_row, file="./symbiodinium_pics/PIC_Sym_output_its2&clade.csv", append=TRUE, sep=",", row.names=FALSE, col.names=FALSE)


## this is a projection of the tree into morphospace
##This code snippit is adapted from a phytools tutorial (http://www.phytools.org/Cordoba2017/ex/3/PICs.html)

pdf(paste(x_trait,"_",y_trait,"_phylomorphospace.pdf",sep=""))
phylomorphospace(tree,cbind(X,Y),xlab=x_trait,ylab=y_trait,label="off",node.size=c(0,0))
points(X,Y,pch=21,bg="grey",cex=1.4)
dev.off()

#Save raw PIC contrasts as a pdf
pdf(paste(x_trait,"_",y_trait,"_pic_scatter_YX.pdf",sep=""))
plot(pic.X,pic.Y,xlab=x_trait,ylab=y_trait,bg='gray',pch=16)
abline(fitYX,col="red")
dev.off()

#Save as a pdf
pdf(paste(x_trait,"_",y_trait,"_pic_scatter_YX_nonzero_PICS_only.pdf",sep=""))
plot(pic.X.nonzero,pic.Y.nonzero,xlab=x_trait,ylab=y_trait,bg='gray',pch=16)
abline(fitYX.nonzero,col="red")
dev.off()

#Save raw data as pdf
pdf(paste(x_trait,"_",y_trait,"_raw_scatter.pdf",sep=""))
plot(X,Y,xlab=x_trait,ylab=y_trait,bg='gray', pch=16)
abline(raw_fitYX,col="red")
dev.off()



#creat a phylogeny heatmap using the phylogenetic tree and alpha diversity metrics
library(phytools)
tree <- "input/symbiodinium_phylogeny_muscle_gblocks_phyml_via_phylogenyfr_09172021.newick"
tree <- read.tree(tree)
tree_table<-sapply()